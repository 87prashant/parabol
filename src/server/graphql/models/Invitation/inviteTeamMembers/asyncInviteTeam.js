import getRethink from 'server/database/rethinkDriver';
import {getUserId} from 'server/utils/authorization';
import getInviterInfoAndTeamName from './getInviterInfoAndTeamName';
import createEmailPromises from './createEmailPromises';
import resolveSentEmails from './resolveSentEmails';
import makeInviteToken from './makeInviteToken';
import makeInvitationsForDB from './makeInvitationsForDB';

export default async function asyncInviteTeam (authToken, teamId, invitees) {
  const r = getRethink();
  const userId = getUserId(authToken);
  const inviteesWithTokens = invitees.map(invitee => ({...invitee, inviteToken: makeInviteToken()}));
  const inviterInfoAndTeamName = await getInviterInfoAndTeamName(teamId, userId);
  const sendEmailPromises = createEmailPromises(inviterInfoAndTeamName, inviteesWithTokens);
  const {inviteesToStore} = await resolveSentEmails(sendEmailPromises, inviteesWithTokens);
  const invitationsForDB = await makeInvitationsForDB(inviteesToStore, teamId, userId);
  // Bulk insert, wait in case something queries the invitation table
  await r.table('Invitation').insert(invitationsForDB);
  return true;
  // TODO generate email to inviter including folks that we couldn't reach
  // if (inviteeErrors.length > 0) {
  // throw errorObj({_error: 'Some invitations were not sent', type: 'inviteSendFail', failedEmails: inviteeErrors});
  // }
};
