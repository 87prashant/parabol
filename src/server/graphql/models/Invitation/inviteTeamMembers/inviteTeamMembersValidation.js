import legitify from 'universal/validation/legitify';
import {fullName, makeInviteeTemplate, task} from 'universal/validation/templates';

export default function inviteTeamMemberValidation(props) {
  const {inviteEmails, activeTeamMemberEmails, pendingApprovalEmails} = props;
  return legitify([{
    email: makeInviteeTemplate(inviteEmails, activeTeamMemberEmails, pendingApprovalEmails),
    fullName,
    task
  }]);
}
