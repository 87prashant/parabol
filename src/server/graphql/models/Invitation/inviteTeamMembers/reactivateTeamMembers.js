// import getRethink from 'server/database/rethinkDriver';
// import getPubSub from 'server/utils/getPubSub';
// import {PRESENCE, REJOIN_TEAM, USER_MEMO} from 'universal/subscriptions/constants';
// import {ADD_TO_TEAM} from 'universal/utils/constants';
// import tmsSignToken from 'server/utils/tmsSignToken';
//
// export default async function reactivateTeamMembers(idsToReactivate, teamId, teamName, exchange, sender) {
//  if (idsToReactivate.length > 0) {
//    const r = getRethink();
//    const userIdsToReactivate = idsToReactivate.map((teamMemberId) => {
//      return teamMemberId.substr(0, teamMemberId.indexOf('::'));
//    });
//    const reactivatedUsers = await r.table('TeamMember')
//      .getAll(r.args(idsToReactivate), {index: 'id'})
//      .update({isNotRemoved: true})
//      .do(() => {
//        return r.table('User')
//          .getAll(r.args(userIdsToReactivate))
//          .update((user) => {
//            return user.merge({
//              tms: user('tms').append(teamId)
//            });
//          }, {returnChanges: true})('changes')
//          .default([])
//          .map((change) => change('new_val'));
//      });
//    reactivatedUsers.forEach((user) => {
//      const {preferredName, id: reactivatedUserId, tms} = user;
//      getPubSub().publish(`${USER_MEMO}/${reactivatedUserId}`, {
//        type: ADD_TO_TEAM,
//        teamId,
//        teamName,
//        _authToken: tmsSignToken({sub: reactivatedUserId}, tms)
//      });
//      const channel = `${PRESENCE}/${teamId}`;
//      exchange.publish(channel, {
//        type: REJOIN_TEAM,
//        name: preferredName,
//        sender
//      });
//    });
//  }
// }
