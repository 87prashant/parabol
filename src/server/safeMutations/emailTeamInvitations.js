import getRethink from 'server/database/rethinkDriver';
import createEmailPromises from '../graphql/models/Invitation/inviteTeamMembers/createEmailPromises';
import makeInvitationsForDB from '../graphql/models/Invitation/inviteTeamMembers/makeInvitationsForDB';
import makeInviteToken from '../graphql/models/Invitation/inviteTeamMembers/makeInviteToken';
import resolveSentEmails from '../graphql/models/Invitation/inviteTeamMembers/resolveSentEmails';

export default async function emailTeamInvitations(invitees, inviter) {
  if (invitees.length === 0) return invitees;
  const {teamId, userId: inviterUserId} = inviter;
  const r = getRethink();
  const inviteesWithTokens = invitees.map((invitee) => ({...invitee, inviteToken: makeInviteToken()}));
  const sendEmailPromises = createEmailPromises(inviter, inviteesWithTokens);
  const {inviteesToStore} = await resolveSentEmails(sendEmailPromises, inviteesWithTokens);
  const invitationsForDB = await makeInvitationsForDB(inviteesToStore, teamId, inviterUserId);
  // Bulk insert, wait in case something queries the invitation table
  return r.table('Invitation').insert(invitationsForDB);
  // TODO generate email to inviter including folks that we couldn't reach
  // if (inviteeErrors.length > 0) {
  // throw errorObj({_error: 'Some invitations were not sent', type: 'inviteSendFail', failedEmails: inviteeErrors});
  // }
};
