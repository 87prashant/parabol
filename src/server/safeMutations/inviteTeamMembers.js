import getRethink from 'server/database/rethinkDriver';
import getResults from 'server/graphql/mutations/helpers/inviteTeamMembers/getResults';
import makeDetailedInvitations from 'server/graphql/mutations/helpers/inviteTeamMembers/makeDetailedInvitations';
import createPendingApprovals from 'server/safeMutations/createPendingApprovals';
import reactivateTeamMembersAndMakeNotifications from 'server/safeMutations/reactivateTeamMembersAndMakeNotifications';
import removeOrgApprovalAndNotification from 'server/safeMutations/removeOrgApprovalAndNotification';
import sendTeamInvitations from 'server/safeMutations/sendTeamInvitations';
import getPendingInvitations from 'server/safeQueries/getPendingInvitations';
import {isBillingLeader} from 'server/utils/authorization';
import publishNotifications from 'server/utils/publishNotifications';
import {ASK_APPROVAL, DENIED, PENDING, REACTIVATE, SEND_INVITATION} from 'server/utils/serverConstants';
import mergeObjectsWithArrValues from 'universal/utils/mergeObjectsWithArrValues';
import resolvePromiseObj from 'universal/utils/resolvePromiseObj';
import approveToOrg from 'server/safeMutations/approveToOrg';

const safeApproveToOrg = async (...args) => {
  let res = {};
  try {
    res = await approveToOrg(...args);
  } catch (e) {
    // noop
  }
  return res;
};

const approvePendingApprovals = async (orgApprovals, inviter) => {
  //
  const {orgId, teamId, userId} = inviter;
  if (!inviter.isBillingLeader) return undefined;
  const otherPendingApprovals = orgApprovals.filter((approval) => approval.status === PENDING && approval.teamId !== teamId);
  const results = await Promise.all(otherPendingApprovals.map((approval) => {
    return safeApproveToOrg(approval.email, orgId, userId);
  }));
  return mergeObjectsWithArrValues(...results);
};

const inviteTeamMembers = async (invitees, teamId, userId) => {
  const r = getRethink();
  const {name: teamName, orgId} = await r.table('Team').get(teamId).pluck('name', 'orgId');

  const emailArr = invitees.map((invitee) => invitee.email);
  const {
    inviterDoc,
    pendingInvitations,
    orgApprovals,
    teamMembers,
    users
  } = await r.expr({
    pendingInvitations: getPendingInvitations(emailArr, teamId)('email')
      .coerceTo('array'),
    orgApprovals: r.table('OrgApproval')
      .getAll(r.args(emailArr), {index: 'email'})
      .filter({orgId, isActive: true})
      .filter((doc) => doc('status').ne(DENIED))
      .coerceTo('array'),
    teamMembers: r.table('TeamMember')
      .getAll(teamId, {index: 'teamId'})
      .coerceTo('array'),
    users: r.table('User')
      .getAll(r.args(emailArr), {index: 'email'})
      .coerceTo('array'),
    inviterDoc: r.table('User').get(userId)
      .pluck('email', 'picture', 'preferredName', 'userOrgs')
  });
  const userOrgDoc = inviterDoc.userOrgs.find((userOrg) => userOrg.id === orgId);
  const inviter = {
    userOrgs: inviterDoc.userOrgs,
    inviterAvatar: inviterDoc.picture,
    inviterEmail: inviterDoc.email,
    inviterName: inviterDoc.preferredName,
    orgId,
    teamId,
    teamName,
    userId,
    isBillingLeader: isBillingLeader(userOrgDoc)
  };
  const detailedInvitations = makeDetailedInvitations(teamMembers, emailArr, users, orgApprovals, pendingInvitations, inviter);
  const inviteesToReactivate = detailedInvitations.filter(({action}) => action === REACTIVATE);
  const inviteesToInvite = detailedInvitations.filter(({action}) => action === SEND_INVITATION);
  const inviteesNeedingApproval = detailedInvitations.filter(({action}) => action === ASK_APPROVAL);

  const pendingApprovalEmails = inviteesNeedingApproval.map(({email}) => email);
  const approvalsToClear = inviteesToInvite.map(({email}) => email);
  const {reactivations, notificationsToClear, teamInvites, newPendingApprovals} = await resolvePromiseObj({
    reactivations: reactivateTeamMembersAndMakeNotifications(inviteesToReactivate, inviter, teamMembers),
    notificationsToClear: removeOrgApprovalAndNotification(orgId, approvalsToClear, {approvedBy: userId}),
    approvePendingApprovals: approvePendingApprovals(orgApprovals, inviter),
    teamInvites: sendTeamInvitations(inviteesToInvite, inviter),
    newPendingApprovals: createPendingApprovals(pendingApprovalEmails, inviter)
  });

  const notificationsToAdd = mergeObjectsWithArrValues(reactivations, teamInvites, newPendingApprovals);
  publishNotifications({notificationsToAdd, notificationsToClear});
  const results = getResults(detailedInvitations);
  return {results};
};

export default inviteTeamMembers;
