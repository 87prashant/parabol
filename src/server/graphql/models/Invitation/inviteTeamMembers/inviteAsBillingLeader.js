import asyncInviteTeam from 'server/safeMutations/asyncInviteTeam';
import removeOrgApprovalAndNotification from 'server/safeMutations/removeOrgApprovalAndNotification';

export default async function inviteAsBillingLeader(invitees, orgId, inviterUserId, teamId) {
  const inviteeEmails = invitees.map((i) => i.email);
  // remove queued approvals
  const promises = [
    removeOrgApprovalAndNotification(orgId, inviteeEmails),
    asyncInviteTeam(inviterUserId, teamId, invitees)
  ];
  await Promise.all(promises);
}
