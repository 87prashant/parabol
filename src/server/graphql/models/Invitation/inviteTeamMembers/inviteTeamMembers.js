import {GraphQLBoolean, GraphQLID, GraphQLList, GraphQLNonNull} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import asyncInviteTeam from 'server/graphql/models/Invitation/inviteTeamMembers/asyncInviteTeam';
import getInviterInfoAndTeamName from 'server/graphql/models/Invitation/inviteTeamMembers/getInviterInfoAndTeamName';
import inviteAsUser from 'server/graphql/models/Invitation/inviteTeamMembers/inviteAsUser';
import {getUserId, getUserOrgDoc, isBillingLeader, requireOrgLeaderOrTeamMember} from 'server/utils/authorization';
import getPubSub from 'server/utils/getPubSub';
import tmsSignToken from 'server/utils/tmsSignToken';
import shortid from 'shortid';
import {ADD_TO_TEAM, PRESENCE, REJOIN_TEAM, USER_MEMO} from 'universal/subscriptions/constants';
import {ALREADY_ON_TEAM, PENDING_APPROVAL, REACTIVATED, SUCCESS, TEAM_INVITE} from 'universal/utils/constants';
import {Invitee} from '../invitationSchema';


//const determineInviteeAction = (invitee) => {
//  const {
//    isActiveTeamMember,
//    isPendingApproval,
//    isPendingInvitation,
//    isUser,
//    isActiveUser,
//    isOrgMember,
//    isNewTeamMember,
//  } = invitee;
//  if (isActiveTeamMember) {
//    return
//  }
//}

const SEND_NOTIFICATION = 'SEND_NOTIFICATION';
const SEND_EMAIL = 'SEND_EMAIL';
const ASK_APPROVAL = 'ASK_APPROVAL';
//const ASK_APPROVAL_EMAIL = 'ASK_APPROVAL_EMAIL';
//const ASK_APPROVAL_EMAIL_NOTIFICATION = 'ASK_APPROVAL_EMAIL_NOTIFICATION';
//const ASK_APPROVAL_NOTIFICATION = 'ASK_APPROVAL_NOTIFICATION';

const sendNotification = async (invitees, inviter) => {
  if (invitees.length === 0) return invitees;
  const r = getRethink();
  const now = new Date();
  const {orgId, inviterName, teamId, teamName} = inviter;
  const invitations = invitees.map((invitee) => ({
    id: shortid.generate(),
    type: TEAM_INVITE,
    startAt: now,
    orgId,
    userIds: [invitee.userId],
    varList: [inviterName, teamId, teamName]
  }));
  await r.table('Notification').insert(invitations);
  return true;
};

const reactivateAndSendWelcomeBack = async (invitees, inviter, exchange) => {
  if (invitees.length === 0) return invitees;
  const {teamId, teamName} = inviter;
  const r = getRethink();
  const userIds = invitees.map(({userId}) => userId);
  const teamMemberIds = userIds.map((userId) => `${userId}::${teamId}`);
  const reactivatedUsers = await r.table('TeamMember')
    .getAll(r.args(teamMemberIds), {index: 'id'})
    .update({isNotRemoved: true})
    .do(() => {
      return r.table('User')
        .getAll(r.args(userIds))
        .update((user) => {
          return user.merge({
            tms: user('tms').append(teamId)
          });
        }, {returnChanges: true})('changes')('new_val')
        .default(null);
    });
  reactivatedUsers.forEach((user) => {
    const {preferredName, id: reactivatedUserId, tms} = user;
    getPubSub().publish(`${USER_MEMO}/${reactivatedUserId}`, {
      type: ADD_TO_TEAM,
      teamId,
      teamName,
      _authToken: tmsSignToken({sub: reactivatedUserId}, tms)
    });
    const channel = `${PRESENCE}/${teamId}`;
    // TODO refactor out exchange
    exchange.publish(channel, {
      type: REJOIN_TEAM,
      name: preferredName,
      sender: inviter.userId
    });
  });
  return true;
}

//const handleInOrgInvite = async (invitee, inviter) => {
//  const {isNewTeamMember, email, userId} = invitee;
//  if (isNewTeamMember) {
//    await sendNotification(userId, inviter);
//    return {
//      result: SUCCESS,
//      email
//    };
//  } else {
//
//    reactivateAndSendWelcomeBack();
//  }
//};

const getAction = (invitee, inviterIsBillingLeader) => {
  const {
    isActiveTeamMember,
    isPendingApproval,
    isPendingInvitation,
    isActiveUser,
    isNewTeamMember,
    isOrgMember,
  } = invitee;
  if (isActiveTeamMember) return ALREADY_ON_TEAM;
  if (isPendingApproval && !inviterIsBillingLeader) return PENDING_APPROVAL;
  if (isPendingInvitation) return SUCCESS;
  if (isOrgMember) {
    return isNewTeamMember ? SEND_NOTIFICATION : REACTIVATED;
  }
  if (inviterIsBillingLeader) {
    return isActiveUser ? SEND_NOTIFICATION : SEND_EMAIL;
  }
  return ASK_APPROVAL;
};

//const getInvitationErrors = (inviteeEmails, activeTeamMembers = [], pendingInvitations = [], pendingApprovals = []) => {
//  const getInviteeError = (inviteeEmail) => {
//    if (activeTeamMembers.includes(inviteeEmail)) {
//      return 'member'
//    }
//    if (pendingInvitations.includes(inviteeEmail)) {
//      return 'invited'
//    }
//    if (pendingApprovals.includes(inviteeEmail)) {
//      return 'pending'
//    }
//    return undefined;
//  };
//
//  return inviteeEmails.map((inviteeEmail) => ({
//    email: inviteeEmail,
//    error: getInviteeError(inviteeEmail)
//  }));
//};

export default {
  type: GraphQLBoolean,
  description: `If in the org,
     Send invitation emails to a list of email addresses, add them to the invitation table.
     Else, send a request to the org leader to get them approval and put them in the OrgApproval table.
     Returns true if invitation is sent, false if approval is needed, undefined/null if neither`,
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The id of the inviting team'
    },
    invitees: {
      type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(Invitee)))
    }
    // notificationId: {
    //   type: GraphQLID
    // }
  },
  async resolve(source, {invitees, teamId}, {authToken, exchange}) {
    const r = getRethink();
    const now = Date.now();

    // AUTH
    await requireOrgLeaderOrTeamMember(authToken, teamId);
    const userId = getUserId(authToken);
    const {name: teamName, orgId} = await r.table('Team').get(teamId).pluck('name', 'orgId');
    const userOrgDoc = await getUserOrgDoc(userId, orgId);
    const inviterIsBillingLeader = isBillingLeader(userOrgDoc);

    // VALIDATION
    // don't let them invite the same person twice
    const emailArr = invitees.map((invitee) => invitee.email);
    const {pendingInvitations, pendingApprovals, teamMembers, users} = await r.expr({
      pendingInvitations: r.table('Invitation')
        .getAll(r.args(emailArr), {index: 'email'})
        .filter((invitation) => invitation('tokenExpiration').ge(r.epochTime(now)))('email')
        .coerceTo('array'),
      pendingApprovals: r.table('OrgApproval')
        .getAll(r.args(emailArr), {index: 'email'})
        .filter({teamId})
        .coerceTo('array'),
      teamMembers: r.table('TeamMember')
        .getAll(teamId, {index: 'teamId'})
        //.merge((teamMember) => ({
        //  userOrgs: r.table('User').get(teamMember('userId'))('userOrgs').default([])
        //}))
        .coerceTo('array'),
      users: r.table('User')
        .getAll(r.args(emailArr), {index: 'email'})
        .coerceTo('array')
    });

    //const newInvitations = getInvitationErrors(emailArr, activeTeamMembers, pendingInvitations, pendingApprovals);

    // RESOLUTION
    const activeTeamMembers = teamMembers.filter((m) => m.isNotRemoved === true).map((m) => m.email);
    const inactiveTeamMembers = teamMembers.filter((m) => !m.isNotRemoved).map((m) => m.email);
    const inviterAndTeamName = await getInviterInfoAndTeamName(teamId, userId);
    const inviterDetails = {
      ...inviterAndTeamName,
      orgId,
      teamId,
      isBillingLeader: inviterIsBillingLeader
    };

    const detailedInvitations = emailArr.map((email) => {
      const userDoc = users.find((user) => user.email === email);
      const details = {
        email,
        isActiveTeamMember: activeTeamMembers.includes(email),
        isPendingApproval: pendingApprovals.includes(email),
        isPendingInvitation: pendingInvitations.includes(email),
        isUser: Boolean(userDoc),
        isActiveUser: userDoc && !userDoc.inactive,
        isOrgMember: userDoc && userDoc.userOrgs.includes((userDocOrg) => userDocOrg.id === orgId),
        isNewTeamMember: !inactiveTeamMembers.includes((tm) => tm.email === email),
        userId: userDoc && userDoc.id
      };
      return {
        ...details,
        action: getAction(details, inviterIsBillingLeader)
      }
    });

    const inviteesToNotify = detailedInvitations.filter(({action}) => action === SEND_NOTIFICATION);
    const inviteesToReactivate = detailedInvitations.filter(({action}) => action === REACTIVATED);
    const inviteesToEmail = detailedInvitations.filter(({action}) => action === SEND_EMAIL);
    const approveWithNotification = detailedInvitations.filter(({action}) => action === ASK_APPROVAL_NOTIFICATION || action === ASK_APPROVAL_EMAIL_NOTIFICATION);
    const approveWithEmail = detailedInvitations.filter(({action}) => action === ASK_APPROVAL_EMAIL || action === ASK_APPROVAL_EMAIL_NOTIFICATION);

    const promises = [
      sendNotification(inviteesToNotify, inviterDetails),
      reactivateAndSendWelcomeBack(inviteesToReactivate, inviterDetails),
      asyncInviteTeam(userId, teamId, inviteesToEmail)


    ]

    const results = await Promise.all(detailedInvitations.map((invitee) => {
      const {
        isActiveTeamMember,
        isPendingApproval,
        isPendingInvitation,
        isUser,
        isActiveUser,
        isOrgMember,
        isNewTeamMember
      } = invitee;

    }));


    //const idsToReactivate = [];
    //const newInvitees = [];
    //for (let i = 0; i < validInvitees.length; i++) {
    //  const validInvitee = validInvitees[i];
    //  const inactiveInvitee = inactiveTeamMembers.find((m) => m.email === validInvitee.email);
    //  if (inactiveInvitee) {
    //    // if they were previously removed from the team, see if they're still in the org
    //    const inOrg = Boolean(inactiveInvitee.userOrgs.find((userOrg) => userOrg.id === orgId));
    //    if (inOrg) {
    //      // if they're in the org, reactive them
    //      idsToReactivate.push(inactiveInvitee.id);
    //      continue;
    //    }
    //  }
    //  newInvitees.push(validInvitee);
    //}
    //
    //await reactivateTeamMembers(idsToReactivate, teamId, teamName, exchange, userId);
    //
    //if (newInvitees.length > 0) {
    //  // if it's a Billing Leader send them all
    //  const inviteeEmails = newInvitees.map((i) => i.email);
    //  if (inviterIsBillingLeader) {
    //    // if any folks were pending, release the floodgates, a Billing Leader has approved them
    //    const pendingApprovals = await removeOrgApprovalAndNotification(orgId, inviteeEmails);
    //    // we should always have a fresh invitee, but we do this safety check in case the front-end validation messes up
    //    const freshInvitees = newInvitees.filter((i) =>
    //      !pendingApprovals.find((d) => d.inviteeEmail === i.email && d.invitedTeamId === teamId));
    //    if (freshInvitees) {
    //      setTimeout(async () => {
    //        await asyncInviteTeam(userId, teamId, freshInvitees);
    //      }, 0);
    //    }
    //    pendingApprovals.forEach((invite) => {
    //      const {inviterId, inviteeEmail, invitedTeamId} = invite;
    //      const invitee = [{email: inviteeEmail}];
    //      // when we invite the person, try to invite from the original requester, if not, Billing Leader
    //      setTimeout(async () => {
    //        await asyncInviteTeam(inviterId, invitedTeamId, invitee);
    //      }, 0);
    //    });
    //
    //    return true;
    //  }
    //  // return false if org approvals sent, true if only invites were sent
    return inviteAsUser(newInvitees, orgId, userId, teamId, teamName);
    //}
    return undefined;
  }
};

