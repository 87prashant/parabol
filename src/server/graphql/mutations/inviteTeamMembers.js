import {GraphQLID, GraphQLList, GraphQLNonNull} from 'graphql';
import Invitee from 'server/graphql/types/Invitee';
import InviteTeamMembersPayload from 'server/graphql/types/InviteTeamMembersPayload';
import inviteTeamMembers from 'server/safeMutations/inviteTeamMembers';
import {getUserId, requireOrgLeaderOrTeamMember} from 'server/utils/authorization';
import publish from 'server/utils/publish';
import {TEAM} from 'universal/subscriptions/constants';
import {ADDED, NOTIFICATION, ORG_APPROVAL, REJOIN_TEAM, REMOVED, TEAM_MEMBER} from 'universal/utils/constants';
import fromTeamMemberId from 'universal/utils/relay/fromTeamMemberId';

export default {
  type: new GraphQLNonNull(InviteTeamMembersPayload),
  description: `If in the org,
     Send invitation emails to a list of email addresses, add them to the invitation table.
     Else, send a request to the org leader to get them approval and put them in the OrgApproval table.`,
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The id of the inviting team'
    },
    invitees: {
      type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(Invitee)))
    }
  },
  async resolve(source, {invitees, teamId}, {authToken, dataLoader, socketId: mutatorId}) {
    const operationId = dataLoader.share();
    // AUTH
    await requireOrgLeaderOrTeamMember(authToken, teamId);
    const viewerId = getUserId(authToken);

    // RESOLUTION
    const subOptions = {mutatorId, operationId};
    const {
      orgApprovalIds,
      reactivations,
      results,
      removedOrgApprovals,
      removedRequestNotifications,
      requestNotifications
    } = await inviteTeamMembers(invitees, teamId, viewerId, subOptions);
    const reactivatedTeamMemberIds = reactivations.map(({teamMemberId}) => teamMemberId);

    // HANDLE REACTIVATION
    reactivations.forEach(({notificationId, teamMemberId, preferredName}) => {
      // send a team member + temporary toast to the rest of the users
      const notification = {type: REJOIN_TEAM, teamId, preferredName};
      publish(TEAM_MEMBER, teamId, ADDED, {teamId, notification}, subOptions);
      // send a team + persisted notification to the reactivated team member
      const {userId} = fromTeamMemberId(teamMemberId);
      publish(TEAM, userId, ADDED, {teamId, notificationId}, subOptions);
    });

    // HANDLE NON-ORG LEADERS NEEDING APPROVAL
    orgApprovalIds.forEach((orgApprovalId) => {
      publish(ORG_APPROVAL, teamId, ADDED, {orgApprovalId}, subOptions);
    });
    requestNotifications.forEach((requestNotification) => {
      const {id: notificationId, userIds} = requestNotification;
      userIds.forEach((userId) => {
        // if the org leader triggers this it won't be in the mutation payload, so mutatorId is not passed in
        publish(NOTIFICATION, userId, ADDED, {notificationId}, {operationId});
      });
    });

    // HANDLE ORG LEADER INVITING A PENDING ORG APPROVAL INVITEE
    const removedOrgApprovalIds = removedOrgApprovals.map(({id}) => id);
    removedOrgApprovalIds.forEach((removedOrgApprovalId) => {
      publish(ORG_APPROVAL, teamId, REMOVED, {removedOrgApprovalId}, subOptions);
    });
    removedRequestNotifications.forEach((notification) => {
      const {userIds} = notification;
      userIds.forEach((userId) => {
        // if the org leader triggers this it won't be in the mutation payload, so mutatorId is not passed in
        publish(NOTIFICATION, userId, REMOVED, {notification}, {operationId});
      });
    });
    return {orgApprovalIds, reactivatedTeamMemberIds, removedOrgApprovalIds, results};
  }
};

