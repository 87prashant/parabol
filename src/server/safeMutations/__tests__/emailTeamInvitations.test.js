import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import makeDataLoader from 'server/__tests__/setup/makeDataLoader';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import newInvitee from 'server/__tests__/utils/newInvitee';
import getRethink from 'server/database/rethinkDriver';
import * as createEmailPromises from 'server/graphql/models/Invitation/inviteTeamMembers/createEmailPromises';
import * as hashInviteTokenKey from 'server/graphql/models/Invitation/inviteTeamMembers/hashInviteTokenKey';
import * as resolveSentEmails from 'server/graphql/models/Invitation/inviteTeamMembers/resolveSentEmails';
import emailTeamInvitations from 'server/safeMutations/emailTeamInvitations';

MockDate.set(__now);
console.error = jest.fn();

describe('emailTeamInvitations', () => {
  test('inserts new records for new invites, updates existing records', async () => {
    // SETUP
    resolveSentEmails.default = jest.fn((_, invitees) => ({inviteesToStore: invitees}));
    createEmailPromises.default = jest.fn();
    hashInviteTokenKey.default = jest.fn(() => Promise.resolve('HA$H'));
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const invitee = newInvitee();
    const invitee2 = newInvitee();
    const dataLoader = makeDataLoader();
    const operationId = dataLoader.share();
    await mockDB.init()
      .newInvitation({email: invitee.email});
    const teamId = mockDB.context.team.id;
    const invitees = [invitee, invitee2];
    const emails = invitees.map(({email}) => email);
    const inviter = {
      teamId,
      userId: mockDB.context.user.id
    };

    // TEST
    await emailTeamInvitations(invitees, inviter, undefined, {operationId});
    // VERIFY
    const db = await fetchAndSerialize({
      invitation: r.table('Invitation').getAll(r.args(emails), {index: 'email'}).orderBy('inviteCount').coerceTo('array')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });
});
