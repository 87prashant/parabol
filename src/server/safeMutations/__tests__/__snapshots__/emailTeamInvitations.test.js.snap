// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emailTeamInvitations inserts new records for new invites, updates existing records 1`] = `
Object {
  "invitation": Array [
    Object {
      "createdAt": 2017-01-01T08:00:00.000Z,
      "email": 2,
      "hashedToken": "HA$H",
      "id": 0,
      "inviteCount": 1,
      "invitedBy": 4,
      "teamId": 5,
      "tokenExpiration": 2017-01-31T08:00:00.000Z,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
    Object {
      "createdAt": 2017-01-01T08:00:00.000Z,
      "email": 3,
      "hashedToken": "HA$H",
      "id": 1,
      "inviteCount": 2,
      "invitedBy": 4,
      "teamId": 5,
      "tokenExpiration": 2017-01-31T08:00:00.000Z,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;
